sklearn/metrics/cluster/tests/test_unsupervised.py

```diff
@@ -169,6 +169,26 @@ def test_non_numpy_labels():
         silhouette_score(list(X), list(y)), silhouette_score(X, y))
 
 
+def test_nonzero_diag_silhouette():
+    # Construct a zero-diagonal matrix
+    dists = pairwise_distances(
+        np.array([[0.2, 0.1, 0.12, 1.34, 1.11, 1.6]]).transpose())
+    # Construct a nonzero-diagonal distance matrix
+    diag_dists = np.diag(np.ones(6)) + dists
+    labels = [0, 0, 0, 1, 1, 1]
+
+    # Test silhouette samples
+    sample_scores = silhouette_samples(dists, labels, metric='precomputed')
+    diag_sample_scores = silhouette_samples(
+        diag_dists, labels, metric='precomputed')
+    assert_array_equal(sample_scores, diag_sample_scores)
+
+    # Test silhoutte score
+    score = silhouette_score(dists, labels, metric='precomputed')
+    diag_score = silhouette_score(diag_dists, labels, metric='precomputed')
+    assert_equal(score, diag_score)
+
```

Comment:
Please check that the sum of dists is unchanged to make sure the user's input is not modified